pyartcd/pyartcd/pipelines/gen_assembly.py

```diff
@@ -0,0 +1,99 @@
+import sys
+
+import click
+from doozerlib import exectools
+from pyartcd import exceptions, github
+from pyartcd.runtime import Runtime
+from pyartcd.cli import cli, click_coroutine, pass_runtime
+
+
+class GenPipeline:
+    def __init__(self, runtime: Runtime, nightlies: str, custom: str, in_flight_prev: str, previous: str,
+                 build_version: str, assembly_name: str, build_number: str):
+        self.runtime = runtime
+        self.nightlies = nightlies
+        self.custom = custom
+        self.in_flight_prev = in_flight_prev
+        self.previous = previous
+        self.build_version = build_version
+        self.assembly_name = assembly_name
+        self.build_number = build_number
+
+    async def run(self):
+        # Create doozer cmd arguments
+        nightly_args = ""
+        for nightly in self.nightlies.split(","):
+            nightly_args += f" --nightly {nightly.strip()}"
+
+        cmd = f"--group openshift-{self.build_version} release:gen-assembly --name {self.assembly_name} from-releases {nightly_args}"
+        if self.custom:
+            cmd += ' --custom'
+            if (self.in_flight_prev and self.in_flight_prev != 'none') or self.previous:
+                print("ERROR: Specifying IN_FLIGHT_PREV or PREVIOUS for a custom release is not allowed.")
+                sys.exit(1)
+
+        else:
+            if self.in_flight_prev and self.in_flight_prev.lower() != 'none':
+                cmd += f" --in-flight {self.in_flight_prev}"
+
+            if self.previous:
+                for previous_data in self.previous.split(','):
+                    cmd += f" --previous {previous_data.strip()}"
+            else:
+                cmd += ' --auto-previous'
+
+        cmd = cmd.replace(" *\\\n *", " ").replace(" *\n *", " ").strip()  # Cleaning whitespace
+
+        # Run doozer command
+        output_all = exectools.cmd_gather(f"doozer --assembly=stream {cmd}")
+        if output_all[0]:
+            print(f"Error: {output_all[0]}")
+            print(output_all[2])
+            sys.exit(1)
+
+        output = output_all[1]
+        release_data = output[output.find("releases:"):].strip()
+        print(release_data)
+
+        base = f"openshift-{self.build_version}"  # The base that we have to branch off of
+        temp_branch = f"automation_{self.build_version}_{self.build_number}"  # format of the temporary branch
+        github_client = github.GithubClient(base=base, data=release_data, assembly_name=self.assembly_name,
+                                            temp_branch=temp_branch)
+
+        try:
+            # Use the output to create a PR
+            if github_client.branch_exists():  # If a branch exists with the same name as the temp branch
+                raise exceptions.BranchExists(
+                    "Temporary branch already exists. Cannot push changes. Delete the temp branch and try again")
+            github_client.sync_with_upstream()  # Sync base branch with upstream
```

Comment:
This seems dangerous. Would like to read releases.yml from the upstream directly, modify it, then push it to the fork.

Reply:
Explained [above](https://github.com/openshift/aos-cd-jobs/pull/3140#discussion_r930329234)